packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -90,3 +92,222 @@ export const useOverflowContainerContext = () => {
+  );
+
+  const isGuessing = React.useRef(false);
+
+  /**
+   * Call this function to guess the new `visibleCount`.
+   * The `visibleCount` is not changed if the correct `visibleCount` has already been found.
+   *
+   * Logic:
+   * - Have a guess range for `visibleCount`. e.g. `[0, 32]` (32 is an arbitrary choice).
+   * - Keep doubling the max guess until the container overflows.
+   *   i.e. the max guess should always be `≥` the correct `visibleCount`.
+   *   - With each such doubling, the new min guess is the current max guess (since underflow = we guessed low).
+   * - Set `visibleCount` to the `maxGuess`.
+   * - Repeat the following by calling `guessVisibleCount()` (keep re-rendering but not painting):
+   *   - Each time the container overflows, new max guess is the average of the two guesses.
+   *   - Each time the container does not overflow, new min guess is the average of the two guesses.
+   * - Stop when the average of the two guesses is the min guess itself. i.e. no more averaging possible.
+   * - The min guess is then the correct `visibleCount`.
+   */
+  const guessVisibleCount = React.useCallback(() => {
+    // If already stabilized, already guessing, or in unit test, do not guess.
+    if (isStabilized || isGuessing.current || isUnitTest) {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      // We need to wait for the ref to be attached so that we can measure available and required sizes.
+      if (containerRef.current == null) {
+        return;
+      }
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (visibleCountGuessRange[1] - visibleCountGuessRange[0] === 1 &&
+          visibleCount === visibleCountGuessRange[0])
+      ) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (visibleCountGuessRange[1] === visibleCount && !isOverflowing) {
+        const doubleOfMaxGuess = visibleCountGuessRange[1] * 2;
+
+        setVisibleCountGuessRange([visibleCount, doubleOfMaxGuess]);
+        setVisibleCount(doubleOfMaxGuess);
+
+        return;
+      }
+
+      let newVisibleCountGuessRange = visibleCountGuessRange;
+
+      if (isOverflowing) {
```

Comment:
I'm wondering if we can simplify this if-else statement into one ternary such as:
```jsx
const newVisibleCountGuessRange = isOverflowing
      ? [visibleCountGuessRange[0], visibleCount]
      : [visibleCount, visibleCountGuessRange[1]];
```